File: base/src/main/java/com/zhangteng/base/base/BaseMvpActivity.java
package com.zhangteng.base.base;

import com.zhangteng.base.mvp.base.BasePresenter;
import com.zhangteng.base.mvp.base.BaseView;

/**
 * 使用Mvp模式Activity基类
 * Created by swing on 2017/11/23.
 */
public abstract class BaseMvpActivity<V extends BaseView, P extends BasePresenter<V>> extends BaseActivity {
    protected P mPresenter;

    @Override
    protected void initView() {
        mPresenter = createPresenter();
        mPresenter.attachView((V) this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mPresenter.detachView();
    }

    /**
     * 子类提供实现
     * 创建对应页面的presenter
     */
    public abstract P createPresenter();
}

File: base/src/main/java/com/zhangteng/base/base/BaseMvpActivity.kt
package com.zhangteng.base.base

import com.zhangteng.base.mvp.base.IModel
import com.zhangteng.base.mvp.base.IPresenter
import com.zhangteng.base.mvp.base.IView

/**
 * 使用Mvp模式Activity基类
 * Created by swing on 2017/11/23.
 */
abstract class BaseMvpActivity<V : IView, M : IModel, P : IPresenter<V, M>> : BaseActivity() {
    protected var mPresenter: P? = null
    override fun initView() {
        mPresenter = createPresenter()
        mPresenter?.attachView(this as V)
        mPresenter?.onStart()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.detachView()
        mPresenter?.onDestroy()
        mPresenter = null
    }

    /**
     * 子类提供实现
     * 创建对应页面的presenter
     */
    abstract fun createPresenter(): P?
}